test/distributed/_composable/test_composability/test_3d_composability.py

```diff
@@ -0,0 +1,757 @@
+        )
+        self._sync_float8_amax_and_scale_history = None
+        #self.compile = job_config.training.compile
+
+    def convert_to_float8_training(self, model: nn.Module):
+        """
+        This function converts the linear layers of `model` to `Float8Linear`.
+        Note that today, only dynamic tensor scaling (the default) is supported.
+        This will mutate the model inplace.
+        """
+        if not self.enabled:
+            return
+
+        from torchao.float8 import convert_to_float8_training
+
+        # Mutates the model inplace replacing instances of nn.Linear with Float8Linear
+        convert_to_float8_training(
+            model,
+            config=self.config,
+            module_filter_fn=lambda mod, fqn: fqn != "output",
+        )
+
+    def precompute_float8_dynamic_scale_for_fsdp(
+        self, model: Union[nn.Module, List[nn.Module]]
+    ):
+        if not self.enabled:
+            return
+
+        if not self.precompute_scale:
+            return
+
+        from torchao.float8 import precompute_float8_dynamic_scale_for_fsdp
+
+        models = [model] if isinstance(model, nn.Module) else model
+        for m in models:
+            precompute_float8_dynamic_scale_for_fsdp(m)
+
+    def sync_float8_amax_and_scale_history(
+        self, model: Union[nn.Module, List[nn.Module]]
+    ):
+        if not self.enabled:
+            return
+
+        if not self.delayed_scaling:
+            return
+
+        from torchao.float8 import sync_float8_amax_and_scale_history
+
+        # TODO(vkuzo): see if precalculating the modules to sync over is going to
+        # meaningfully help performance
+
+        if self._sync_float8_amax_and_scale_history is None:
+            if self.compile:
+                self._sync_float8_amax_and_scale_history = torch.compile(
+                    sync_float8_amax_and_scale_history
+                )
+            else:
+                self._sync_float8_amax_and_scale_history = (
+                    sync_float8_amax_and_scale_history
+                )
+
+        models = [model] if isinstance(model, nn.Module) else model
+        for m in models:
+            self._sync_float8_amax_and_scale_history(m)
+
+
+class Test3DComposability(FSDPTest):
+    logger = logging.getLogger()
+    logger.info("Starting job: 3D composability test")
+    DeviceType = Union[int, str, torch.device]
+
+    def _init_logger(self):
+        self.logger.setLevel(logging.INFO)
+        ch = logging.StreamHandler()
+        ch.setLevel(logging.INFO)
+        formatter = logging.Formatter(
+            "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+        )
+        ch.setFormatter(formatter)
+        self.logger.addHandler(ch)
+
+        os.environ["KINETO_LOG_LEVEL"] = "5"
+
+    @skip_if_lt_x_gpu(4)
+    def test_3d_composability(self):
+        self.run_subtests(
+            {
+                "dp_degree": [
+                    1,
```

Comment:
dp_degree = 1 means disable FSDP?

Reply:
later would change the subset structure to cover more tests
Currently could run with dp, tp, pp degree in [1,2,4]

Reply:
it's just I remember dp=1 means no dp. so for model init, wnat to make sure fully_shard is indeed called in current setting